src/colmap/util/logging.h

```diff
@@ -61,6 +63,66 @@
 #define CHECK_OPTION_GE(val1, val2) CHECK_OPTION_OP(_GE, >=, val1, val2)
 #define CHECK_OPTION_GT(val1, val2) CHECK_OPTION_OP(_GT, >, val1, val2)
 
+// Option checker macros. In contrast to glog, this function does not abort the
+// program, but simply throws an exception on failure.
+#define THROW_EXCEPTION(exception, msg) \
+  throw TemplateException<exception>(__FILE__, __LINE__, ToString(msg));
+
+#define THROW_CUSTOM_CHECK_MSG(condition, exception, msg) \
+  if (!(condition))                                       \
+    throw TemplateException<exception>(                   \
+        __FILE__,                                         \
+        __LINE__,                                         \
+        __GetCheckString(#condition) + std::string(" ") + ToString(msg));
+
+#define THROW_CUSTOM_CHECK(condition, exception) \
+  if (!(condition))                              \
+    throw TemplateException<exception>(          \
+        __FILE__, __LINE__, __GetCheckString(#condition));
+
+#define THROW_CHECK(expr) __ThrowCheckImpl(__FILE__, __LINE__, !!(expr), #expr);
+
+#define THROW_CHECK_MSG(expr, msg) \
+  __ThrowCheckImplMsg(__FILE__, __LINE__, !!(expr), #expr, ToString(msg));
+
+#define THROW_CHECK_NOTNULL(val) \
+  __ThrowCheckNotNull(__FILE__, __LINE__, (val), #val)
+
+#define THROW_CHECK_OP(name, op, val1, val2) \
+  __ThrowCheckOpImpl(                        \
+      __FILE__, __LINE__, (val1 op val2), val1, val2, #val1, #val2, #op);
+
+#define THROW_CHECK_EQ(val1, val2) THROW_CHECK_OP(_EQ, ==, val1, val2)
+#define THROW_CHECK_NE(val1, val2) THROW_CHECK_OP(_NE, !=, val1, val2)
+#define THROW_CHECK_LE(val1, val2) THROW_CHECK_OP(_LE, <=, val1, val2)
+#define THROW_CHECK_LT(val1, val2) THROW_CHECK_OP(_LT, <, val1, val2)
+#define THROW_CHECK_GE(val1, val2) THROW_CHECK_OP(_GE, >=, val1, val2)
+#define THROW_CHECK_GT(val1, val2) THROW_CHECK_OP(_GT, >, val1, val2)
+
+#define THROW_CHECK_OP_MSG(op, val1, val2, msg) \
+  __ThrowCheckOpImpl(__FILE__,                  \
+                     __LINE__,                  \
+                     (val1 op val2),            \
+                     val1,                      \
+                     val2,                      \
+                     #val1,                     \
+                     #val2,                     \
+                     #op,                       \
+                     ToString(msg));
+
+#define THROW_CHECK_EQ_MSG(val1, val2, msg) \
+  THROW_CHECK_OP_MSG(==, val1, val2, msg)
+#define THROW_CHECK_NE_MSG(val1, val2, msg) \
+  THROW_CHECK_OP_MSG(!=, val1, val2, msg)
+#define THROW_CHECK_LE_MSG(val1, val2, msg) \
+  THROW_CHECK_OP_MSG(<=, val1, val2, msg)
+#define THROW_CHECK_LT_MSG(val1, val2, msg) \
+  THROW_CHECK_OP_MSG(<, val1, val2, msg)
+#define THROW_CHECK_GE_MSG(val1, val2, msg) \
+  THROW_CHECK_OP_MSG(>=, val1, val2, msg)
+#define THROW_CHECK_GT_MSG(val1, val2, msg) \
+  THROW_CHECK_OP_MSG(>, val1, val2, msg)
```

Comment:
I'm no too happy with these duplicated macros but I couldn't find a clean way to provide macro overload. Piping messages like glog with the `<<` syntax would be nicer but requires throwing in the destructor - I'm not convinced that we can easily make it safe.
https://github.com/google/glog/blob/v0.7.0-rc1/src/glog/logging.h#L1217-L1220
cc @Phil26AT